src/metadata.rs

```diff
@@ -0,0 +1,162 @@
+use std::collections::HashMap;
+
+use crate::{
+    account::{account, Account},
+    amount::{amount, currency, Amount},
+    date::{date, Date},
+    string::string,
+};
+
+use nom::{
+    branch::alt,
+    bytes::complete::tag,
+    character::complete::{line_ending, one_of, satisfy, space0, space1},
+    combinator::{map, recognize},
+    multi::{fold_many0, many0},
+    sequence::{pair, preceded, separated_pair, tuple},
+    IResult,
+};
+
+pub type Metadata<'a> = HashMap<String, Value<'a>>;
+
+#[derive(Clone, Debug, PartialEq)]
+#[non_exhaustive]
+pub enum Value<'a> {
+    Account(Account<'a>),
+    Amount(Amount<'a>),
+    Currency(String),
+    Date(Date),
+    String(String),
```

Comment:
Why use `String` instead of `&'a str`?

Reply:
I'm very much a Rust newbie, so please forgive my naïveté. If I use `String(&'a str)`, I'm not sure how to construct the `String`, because here `map(string, |s| Value::String(&s))`, I'll be borrowing `s` defined only in the local function. I think this is the result of the `string` parser in `string.rs` defined as `fn string(input: &str) -> IResult<&str, String>`.

I'll do some more reading to better understand how these work.

Reply:
in theory, if you don't transform the input string it should be possible to use `&'a str`, where `'a` is the lifetime of the input string given to the parser. For example, I could use string slices for, the account names and amount currency.

But it is not always possible though. For example, the transaction description is transformed, ("\n" is replaced by a new line), therefore it is impossible to reference a slice in user-input. A new string must be allocated.

I am not 100% sure if `&str` is possible here, but it feels like it's worth looking at it. I'll try to look at it more closely soon.

For the time being, if you don't see how to make it `&str`, you may gate the whole feature behind the `#[cfg(unstable)]`, and I can merge it as a new "unstable" feature, so that we can continue to tweak and change the API after merge.